### ISSUE TITLE
GIT supportsSubmoduleCheckout prevents to update submodule

### ISSUE THREAD

**Issue by tardyp** [buildbot/buildbot#3028]
This ticket is a migrated Trac ticket [3627](http://trac.buildbot.net/ticket/3627)

People contributed to the original ticket: @tardyp, @unknown_contributor
Ticket created on: `Oct 18 2016`
Ticket last modified on: `Feb 08 2017`

---

The commit [[a310e30afab77fd6b04c9f728bf73ff7d8f4d847|https://github.com/buildbot/buildbot/commit/a310e30afab77fd6b04c9f728bf73ff7d8f4d847]] changed the default behavior for 'submodule update'.
From my point of view this is wrong and should not be the default value. Overriding the checkout policy may be a useful parameter to be added to the GIT constructor.

The GIT documentation says the following for "--checkout":
*This is the default behavior, the main use of this option is to override submodule.$name.update when set to a value other than checkout. If the key submodule.$name.update is either not explicitly set or set to checkout, this option is implicit.*

This means the only purpose for '--checkout' is to override the update policy to checkout mode, this is already GIT's default. There is technically no need to override the GIT standard.

We have a large project with several submodules, where the .gitmodules is configured as follows:

    [["sub_name"|submodule]]
      path = name.git
      update = none

Our buildbot user (and most of our userbase) do not have the access rights  to this repository. Therefore the default policy is update = none.
This worked correctly with 0.8.12. Overriding the update policy leads to checkout failure, due to missing access rights.

In addition to that there is no possibility to oversteer this behavior, as this is determined by the GIT version only.

---

**Comment by krf** [comment]
I just ran into this issue and I agree with the reporter. `--checkout` shouldn't be passed to `git submodule update`.

Consider the use case of disabling a specific submodule since the CI worker does not have (and never will get) access to a certain remote site.

Examle: Say `myCustomersSubmodule` is an inaccessible submodule and we just want to disable it via Git config:

  `git -c submodule.myCustomersSubmodule.update=none submodule update --init --force --recursive --checkout` *still* tries to access `myCustomersSubmodule.url`

... does NOT work: since `--checkout` overwrites the update method, regardless.

`git -c submodule.myCustomersSubmodule.update=none submodule update --init --force --recursive` *still* tries to access `myCustomersSubmodule.url`

... (without `--checkout`) does work.

**Comment by krf** [comment]
@wentasah You're the original author of the commit. Comments?

**Comment by krf** [comment]
Here's the revert commit, fwiw: https://github.com/KDAB/buildbot/commit/f3515c1a03326520e97d31b0c76ed10ff2f3fbff

**Comment by wentasah** [comment]
I agree with @tardyp that `--checkout` should not be forced, but it should be possible to specify it in the constructor.

My original need for having `--checkout` was the following: We had `update = merge` in .gitmodules and when buildbot needed to build a different branch or an older commit, the result of merge in the submodule was wrong. In other words, the submodule commit used for the build depended on the build history rather than solely on the commit being built.
